### DISCUSSION TITLE
[OP Stack Deployment] How can I transfer ETH from L1 to my own L2 to be used as gas?

### DISCUSSION THREAD

**Discussion by yhg1225** [ethereum-optimism/developers#685] [Deploying the OP Stack] [answered]
### Did you check the documentation?

- [X] Yes

### Did you check for duplicate questions?

- [X] Yes

### Issue Description

Hey team, I need some development support here.

I am running my own custom L2 chain using the OP stack, and I followed this tutorial:
https://docs.optimism.io/builders/chain-operators/tutorials/create-l2-rollup

I am running my custom L2 with the Holesky L1 testnet locally. So far, everything is up and running smoothly, but I can't create and submit transactions on my custom L2.

I tried to execute the final steps and send some Sepolia ETH to the L1StandardBridgeProxy contract, but I can't receive it on my own L2.

How can I transfer ETH from L1 to my own L2 to be used as gas?

Please help me resolve this issue.

### Logs

op-geth log
![op-geth](https://github.com/user-attachments/assets/0da817a6-68d2-48a9-a41c-732ec59db2e4)
op-node log
![op-node](https://github.com/user-attachments/assets/fab014fc-e704-43dc-8936-fa424062c297)
op-batcher log
![op-batcher](https://github.com/user-attachments/assets/164d9354-1bbe-4f17-93dd-07d2641f5c41)
op-proposer
![op-proposer](https://github.com/user-attachments/assets/f7a4bef7-5946-4bc1-a13c-cbefabdc4d1b)

### Additional Information

Hey team, I need some development support here.

I am running my own custom L2 chain using the OP stack, and I followed this tutorial for depositing ETH:
https://docs.optimism.io/builders/app-developers/tutorials/cross-dom-bridge-eth#deposit-eth

Steps one to five were completed without any issues, but I encountered an error at the final step.

Please help me resolve this issue.
<img width="1062" alt="image" src="https://github.com/user-attachments/assets/22d5eda1-90bf-4fbd-bc07-66d25059b6d7" />

**Comment by Chomtana** [comment]
Logs looks normal

Have you check if you have passed correct L1 / L2 chain for public and wallet client and L1StandardBridgeProxy address for your chain in the process of depositing ETH. If you are not setting them correctly, it's likely to be deposited to into Optimism Sepolia instead of your chain.

Can you show the L1 tx hash (`depositHash`) from step 3

For reference, you should update your `portalAddress` in step 3 according to this docs: https://viem.sh/op-stack/actions/depositTransaction#portaladdress-optional

Or you need to define your custom OP stack chain correctly with `targetChain.contracts.portal[chainId].address` as your OptimismPortalProxy contract

And you also need to make sure public and wallet client connect to the correct chain.

**Reply by yhg1225** [reply]
Thx for your reply
1.I executed the following command to retrieve the address of the L1StandardBridgeProxy contract:
`cat deployments/getting-started/.deploy | jq -r .L1StandardBridgeProxy`
2.I obtained the address of the L1StandardBridgeProxy contract: 0xE25B3D82Ef4f7e8A4eDbBD4D20E60AFf04FdABF7.
3.I sent some Sepolia ETH to the L1StandardBridgeProxy contract, but I was unable to see the ETH on L2.
4.I attempted to verify the balances using the following Python code:
```
import asyncio
from web3 import AsyncWeb3

async def test():
    w3 = AsyncWeb3(AsyncWeb3.AsyncHTTPProvider('http://localhost:8545'))
    result = await w3.is_connected()
    print(result)

    balance1 = await w3.eth.get_balance("0x2094E02825BA403cd305Ae8f45F0e2b668b63B39")
    balance2 = await w3.eth.get_balance("0x2b4F81D922Df403d7aF880BD0e647c7bF9c9dd4A")
    balance3 = await w3.eth.get_balance("0x8cbDeA6Ac30B5FB5Cba5eC3190ea2940ef1BFE76")
    print(balance1) 
    print(balance2)
    print(balance3)

loop = asyncio.get_event_loop()
loop.run_until_complete(test())
```

The results were:
```
True
0
0
0
```

Based on all the steps, I believe the public client and wallet client should be connected to the correct chain.

Additionally, the L1 transaction hash from step 3 is 0x843aceab650f9b9f82fc79e1c2655772d02deeacee127b5aeaefc3450e23462f.

I will try your suggestions and reply here.
looking forward to your further response.

**Reply by yhg1225** [reply]
I tried again,but faild again
1.
`const depositArgs = await publicClientL2.buildDepositTransaction({
    mint: parseEther("0.0001"),
    to: account.address,
});`
![WechatIMG398](https://github.com/user-attachments/assets/f0a41b2b-b429-4255-9b77-9009a64fa81b)
2.
`const depositReceipt = await publicClientL1.waitForTransactionReceipt({ hash: depositHash });
console.log('L1 transaction confirmed:', depositReceipt);`
<img width="1044" alt="WechatIMG400 1" src="https://github.com/user-attachments/assets/8ca8e20e-4891-4e26-84be-021463327dc5" />
3.
```
const [l2hash] = getL2TransactionHashes(depositReceipt);
# l2hash: "0x41416957b44c7ce5c1622a0f8b3af412a6f1e9148c20e89a699100cf872baf56" 
const l2Receipt = await publicClientL2.waitForTransactionReceipt({
    hash: l2hash,
 }); 
```
<img width="1055" alt="image" src="https://github.com/user-attachments/assets/8082bb78-82ff-45f9-a814-37a3c8be256a" />

**Reply by Chomtana** [reply]
You are sending the transaction in Sepolia but you have mentioned that your chain is settled on Holesky

https://sepolia.etherscan.io/tx/0x843aceab650f9b9f82fc79e1c2655772d02deeacee127b5aeaefc3450e23462f

And the L1StandardBridge you are sending to is for OP Sepolia OptimismPortalProxy: https://sepolia.etherscan.io/address/0x16fc5058f25648194471939df75cf27a2fdc48bc

As seen in https://docs.optimism.io/chain/addresses

**Reply by yhg1225** [reply]
It's my mistake. I changed the OptimismPortalProxy contract to my own:
`depositArgs.targetChain.contracts.portal["11155111"].address = "0xE25B3D82Ef4f7e8A4eDbBD4D20E60AFf04FdABF7"`
The L1 transaction: https://sepolia.etherscan.io/tx/0xd50199f86e77bc74462219d22946c8cba3d911e2eccd70c1f3dcf6a32209d68b
the error is:
<img width="1237" alt="image" src="https://github.com/user-attachments/assets/bf736424-b61a-4ad6-abbd-dc964dc711b2" />

**Reply by Chomtana** [reply, marked as answer]
It's sending to wrong address, it send to 0x2A39F5A0B2Dc87aC4dbB966B69fe9b1Debe215B5 instead

You should build `publicClientL2` with correct chain definition at start so that you don't need to set parameter in an unusal way

```typescript
import { createPublicClient, http } from 'viem';
import { defineChain } from 'viem'
import { chainConfig } from 'viem/op-stack'

const sourceId = 11_155_111 // sepolia

export const yourchain = /*#__PURE__*/ defineChain({
  ...chainConfig,
  id:<YOUR_CHAIN_ID>,
  name: '<YOUR_CHAIN_NAME>',
  nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['<YOUR_RPC_URL>'],
    },
  },
  blockExplorers: {
    default: {
      name: 'Blockscout',
      url: '<YOUR_EXPLORER_URL>',
      apiUrl: '<YOUR_EXPLORER_URL>/api',
    },
  },
  contracts: {
    ...chainConfig.contracts,
    disputeGameFactory: {
      [sourceId]: {
        address: '<ADDRESS>',
      },
    },
    l2OutputOracle: {
      [sourceId]: {
        address: '<ADDRESS>',
      },
    },
    multicall3: {
      address: '0xca11bde05977b3631167028862be2a173976ca11',
      blockCreated: 0,
    },
    portal: {
      [sourceId]: {
        address: '<ADDRESS>',
        blockCreated: <DEPLOYED_BLOCK>,
      },
    },
    l1StandardBridge: {
      [sourceId]: {
        address: '<ADDRESS>',
        blockCreated: <DEPLOYED_BLOCK>,
      },
    },
  },
  sourceId,
})

const publicClientL2 = createPublicClient({
  chain: yourchain,
  transport: http(),
});
```

**Reply by yhg1225** [reply]
**my .deploy config**
```
{
  "AddressManager": "0x3b89baE801E50c58Da1079a6012351877CCfbe7c",
  "DisputeGameFactory": "0x9Efddb10f1995eA710C305f955Ff74890B083F00",
  "DisputeGameFactoryProxy": "0xA2c52d9d09F5E52922b249ebBD217fDDd6588691",
  "L1CrossDomainMessenger": "0xdfFc1e9ACa8cb4c5ca3b41Ec532af2926E5818b3",
  "L1CrossDomainMessengerProxy": "0xE950ae6F8EFB1710D8982388C93848Ee58f22fBb",
  "L1ERC721Bridge": "0xDDa0F0E220CbbA4b8B2F676Fa8D4E449bE49e2a6",
  "L1ERC721BridgeProxy": "0xb51c895053ef349bD329d0bf34E7638D0Be176DB",
  "L1StandardBridge": "0x06B5E5174e3721D1a5c22Ff03ce8b73DB544eea2",
  "L1StandardBridgeProxy": "0xE25B3D82Ef4f7e8A4eDbBD4D20E60AFf04FdABF7",
  "L2OutputOracle": "0x8775166F839754a724B158094747843Bb69E40AE",
  "L2OutputOracleProxy": "0x313EF13ec2d4b393522C648c0e4c3650B297AF7A",
  "Mips": "0x6350dD628B064Db5F1A6DDA33F3b4AB25BC510C8",
  "OptimismMintableERC20Factory": "0xA8EB6b29Fa5caB14d374B3884A17Ea33c9c613E1",
  "OptimismMintableERC20FactoryProxy": "0x66508DB135fFb0B96D26e45F46b0ADc824E586E1",
  "OptimismPortal": "0x3454dEAdE69AE0A42F227d574761B4b35184923b",
  "OptimismPortalProxy": "0x2A39F5A0B2Dc87aC4dbB966B69fe9b1Debe215B5",
  "PreimageOracle": "0xcFe3B4ffA70Ffa15b3F9e69F7dba898E4321a1dF",
  "ProtocolVersions": "0x399DA7AE72c67Cf2E82AD4F021c151a2E2D94e6d",
  "ProtocolVersionsProxy": "0x005e110D7907C51419edE715A7Bf5b1E1118EFd6",
  "ProxyAdmin": "0x8d1304268f68a0A7262e24B2208f73bABF49C0ed",
  "SafeProxyFactory": "0xa6B71E26C5e0845f74c812102Ca7114b6a896AB2",
  "SafeSingleton": "0xd9Db270c1B5E3Bd161E8c8503c55cEABeE709552",
  "SuperchainConfig": "0x93c0172F8F92578b7923BAF945B8D29447728436",
  "SuperchainConfigProxy": "0x4234158cb3dCBC8d0719d7D7E5f294446369e6Fa",
  "SystemConfig": "0xf5Bdb7375729eaAe8b7f5920F8B3D489F7742C77",
  "SystemConfigProxy": "0x2ed4a4032DffeD414E96Ede77c2eE8922A2D9c38",
  "SystemOwnerSafe": "0xEaccae38A88b4aE46AeA666AFA8FFDFBcAcFD8EB"
}
```
**my demo**
```javascript
// require('dotenv').config();
import { createPublicClient, http, createWalletClient, parseEther, formatEther } from 'viem';
import { sepolia, optimismSepolia } from 'viem/chains';
import { privateKeyToAccount } from 'viem/accounts';
import { getL2TransactionHashes, publicActionsL1, publicActionsL2, walletActionsL1, walletActionsL2 } from 'viem/op-stack';
import { chainConfig } from "viem/op-stack"
import { defineChain } from "viem/utils"


const PRIVATE_KEY = "0x13435e986649057c2c0165e09c9dfe87d085";
const account = privateKeyToAccount(PRIVATE_KEY);


const sourceId = 11_155_111 // sepolia

export const MySepolia = /*#__PURE__*/ defineChain({
  ...chainConfig,
  id: 42069,
  name: 'pegglecoin',
  nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://localhost:8545'],
    },
  },
  contracts: {
    ...chainConfig.contracts,
    disputeGameFactory: {
      [sourceId]: {
        address: '0x9Efddb10f1995eA710C305f955Ff74890B083F00',
      },
    },
    l2OutputOracle: {
      [sourceId]: {
        address: '0x8775166F839754a724B158094747843Bb69E40AE',
      },
    },
    multicall3: {
      address: '0xca11bde05977b3631167028862be2a173976ca11',
      blockCreated: 0,
    },
    portal: {
      [sourceId]: {
        address: '0x2A39F5A0B2Dc87aC4dbB966B69fe9b1Debe215B5',
      },
    },
    l1StandardBridge: {
      [sourceId]: {
        address: '0x06B5E5174e3721D1a5c22Ff03ce8b73DB544eea2',
      },
    },
  },
  testnet: true,
  sourceId,
})

const publicClientL1 = createPublicClient({
    chain: sepolia,
    transport: http("https://eth-sepolia.g.alchemy.com/v2/lpG7lK-fo6x9vt_c2Y-moe2u5703pAR4"),
}).extend(publicActionsL1());

const walletClientL1 = createWalletClient({
    account,
    chain: sepolia,
    transport: http("https://eth-sepolia.g.alchemy.com/v2/lpG7lK-fo6x9vt_c2Y-moe2u5703pAR4"),
}).extend(walletActionsL1());



const publicClientL2 = createPublicClient({
    chain: MySepolia,
    transport: http("http://localhost:8545"),
}).extend(publicActionsL2());

const walletClientL2 = createWalletClient({
    account,
    chain: MySepolia,
    transport: http("http://localhost:8545"),
}).extend(walletActionsL2());

async function main() {
    try {
        // const l1Balance = await publicClientL1.getBalance({ address: account.address });
        // console.log(`L1 Balance: ${formatEther(l1Balance)} ETH`);

        await depositETH();
    } catch (error) {
        console.error('Error in main function:', error);
    }
}

async function depositETH() {
    try {
        const depositArgs = await publicClientL2.buildDepositTransaction({
            mint: parseEther("0.0001"),
            to: account.address,
        });
        console.log(`Deposit transaction hash on L1: ${depositArgs}`);
        
        const depositHash = await walletClientL1.depositTransaction(depositArgs);
        console.log(`Deposit transaction hash on L1: ${depositHash}`);

        const depositReceipt = await publicClientL1.waitForTransactionReceipt({ hash: depositHash });
        console.log('L1 transaction confirmed:', depositReceipt);

        const [l2Hash] = getL2TransactionHashes(depositReceipt);
        console.log(`Corresponding L2 transaction hash: ${l2Hash}`);

        const l2Receipt = await publicClientL2.waitForTransactionReceipt({
            hash: l2Hash,
        });
        console.log('L2 transaction confirmed:', l2Receipt);
        console.log('Deposit completed successfully!');
    } catch (error) {
        console.error('Error in depositETH function:', error);
    }
}

main();

``` 

Here is my configuration file and demo. Could you help me identify where the issue might be?
My branch is tutorials/chain. Should I switch to a different branch for deployment?

**Reply by opfocus** [reply]
```
  contracts: {
    ...chainConfig.contracts,
    disputeGameFactory: {
```
note that use proxy contracts here, not the implement contracts address

**Comment by opfocus** [comment]
_I tried to execute the final steps and send some Sepolia ETH to the L1StandardBridgeProxy contract, but I can't receive it on my own L2._

This is because your L2 has not completed the sync, you will see it on the L2 when the sync is complete
